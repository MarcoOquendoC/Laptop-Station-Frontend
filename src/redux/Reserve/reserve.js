import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import api from '../../api/api';

const GET_RESERVES = 'GET_RESERVES';
// const POST_RESERVE = 'POST_RESERVE';
const DELETE_RESERVE = 'DELETE_RESERVE';

const initialState = [];

// Thunk
const getReservesInfo = createAsyncThunk(GET_RESERVES, async () => {
  try {
    return await api.fetchReserves();
  } catch (error) {
    return error.message;
  }
});

const deleteReserve = createAsyncThunk(DELETE_RESERVE, async (id) => {
  try {
    return await api.deleteReservefetch(id);
  } catch (error) {
    return error.message;
  }
});

// reducer
export const itemSlice = createSlice({
  name: 'reserves',
  initialState,
  reducers: { },
  extraReducers: {
    [getReservesInfo.fulfilled]: (state, action) => action.payload,
    [deleteReserve.fulfilled]: (state, action) => action.payload,
  },
});

export { getReservesInfo, deleteReserve };
export default itemSlice.reducer;
