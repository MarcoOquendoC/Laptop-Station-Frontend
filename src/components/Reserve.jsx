import React, { useState, useEffect } from 'react';
import {
  Link, useLocation, useParams, useNavigate,
} from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { getReservesInfo, addReserve, deleteReserve } from '../redux/Reserve/reserve';
import NavPrincipal from './NavPrincipal';
import arrowLeft from '../images/left.svg';

const Reserve = () => {
  const params = useParams();
  const { itemId } = params;
  const navigate = useNavigate();
  const auth = useSelector((store) => store.auth);
  const reserves = useSelector((store) => store.reserves);

  const [state, updateState] = useState('');
  const [date, setDate] = useState('2023-12-14');
  const location = useLocation();

  const handleBack = () => {
    if (location.pathname === '/details') {
      updateState(true);
    } else {
      updateState(!state);
    }
  };

  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getReservesInfo());
  }, [dispatch]);

  const addReservation = () => {
    const reserve = {
      date,
      user_id: auth.id,
      item_id: itemId,
    };
    dispatch(addReserve(reserve));
    navigate('/reservations');
  };

  const handleDelete = (reserveId) => {
    dispatch(deleteReserve(reserveId));
  };

  return (
    <>
      <NavPrincipal />
      <div className="viewfinder">
        <Link to={`/detail/${itemId}`} onClick={() => handleBack()} className="btn left"><img src={arrowLeft} alt="left" /></Link>
        {
          auth.first_name ? (
            <div>
              <h1>
                New Reservation
              </h1>
              <form onSubmit={addReservation}>
                <input className="date" onChange={(e) => setDate(e.target.value)} type="date" placeholder="2000-12-31" />
                <button type="submit">Add</button>
              </form>
              <div className="reserves">
                <h2>Reservations for this laptop</h2>
                {
                  reserves.filter((r) => r.item_id === Number(itemId)).map((reserve) => (
                    <div key={reserve.id}>
                      <strong>{reserve.title}</strong>
                      {': '}
                      <span>{reserve.date}</span>
                      <button type="button" onClick={() => handleDelete(reserve.id)}>ðŸ—‘</button>
                    </div>
                  ))
                }
              </div>
            </div>
          ) : <div className="msg_alert">Please login or register to manage Reserves</div>
        }
      </div>
    </>
  );
};

export default Reserve;
