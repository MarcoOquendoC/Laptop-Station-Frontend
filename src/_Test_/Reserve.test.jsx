import '@testing-library/jest-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { screen, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import Reserve from '../components/Reserve';
import store from '../redux/configureStore';

describe('Testing Reserve', () => {
  let reserve;
  beforeAll(async () => {
    reserve = renderer
      .create(
        <Provider store={store}>
          <Router>
            <Reserve />
          </Router>
        </Provider>,
      )
      .toJSON();
  });

  it('should renders correctly', () => {
    expect(reserve).toBeTruthy();
  });

  it('should match the snapshot', () => {
    expect(reserve).toMatchSnapshot();
  });

  test('renders the component with the correct heading', () => {
    const heading = waitFor(() => screen.getByRole('heading', { name: 'Reserve' }));
    waitFor(() => expect(heading).toBeInTheDocument());
  });
});
