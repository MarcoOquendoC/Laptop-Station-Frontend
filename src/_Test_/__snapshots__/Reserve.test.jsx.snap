// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Reserve should match the snapshot 1`] = `
Array [
  <a
    href="/detail/undefined"
    onClick={[Function]}
  >
    Back
  </a>,
  <div>
    <h1>
      Hola reserva
    </h1>
    <form
      action=""
    >
      <input
        type="date"
      />
      <input
        type="number"
      />
      <input
        type="number"
      />
    </form>
    <div
      className="reserves"
    >
      <div
        className="reserve"
      >
        <h2>
          Este es el user_id:
           
        </h2>
        <p>
          Fecha:
           
        </p>
        <p>
          laptop: Nombre del item
        </p>
      </div>
    </div>
  </div>,
]
`;
